alphazero/mcts/mcts.py

```diff
@@ -0,0 +1,282 @@
+        game (GameClassType): Game that is played
+        rollout (bool, optional): Whether the search uses rollout instead of a network. Defaults to False.
+        net (torch.nn.Module, optional): Needs to be input if rollout = false
+        n_iter (int, optional): How many iterations to run Defaults to 1600.
+        c_puct (float, optional): c_puct value for node evaluation. Defaults to 1.0.
+        temperature (float, optional): Temperature for policy calculation. Defaults to 1.0.
+        eval (str, optional): Evaluation type. Defaults to "PUCT".
+    Returns:
+        policy (torch.Tensor)
+        sampled_node (MCTSNode)
+    """
+
+    def backpropagate(leaf_node: MCTSNode):
+        """
+        Backpropagates up the tree after reaching leaf node
+        Args:
+            leaf_node (MCTSNode): Leaf node to backpropagate from
+        """
+        if not rollout:
+            _, v_leaf = leaf_node.calc_policy_value(
+                rollout=rollout, network=net
+            )  # Think about batching/efficiency
+        else:
+            _, v_leaf = leaf_node.calc_policy_value(rollout=rollout, game=game)
+
+        leaf_node.is_terminal = game.is_game_over(leaf_node.state)
+        if not leaf_node.is_terminal:
+            leaf_node.expand(game)
+        current = leaf_node
+        while not current.is_root:
+            current.n_visit += 1
+            current.total_value += v_leaf
+            current = current.parent
+
+    def forward(root_node: MCTSNode):
+        """
+        Runs a forward pass through the tree and then backpropagates back through the tree
+        Args:
+            root_node (MCTSNode): Root node to start from
+        """
+        current = root_node
+        while not current.is_leaf and not current.is_terminal:
+            sum_n = np.sum([child.n_visit for child in current.children])
+            current = current.children[
+                np.argmax(
+                    [
+                        child.confidence_bound(sum_n=sum_n, c_puct=c_puct)
+                        for child in current.children
+                    ]
+                )
+            ]
+        backpropagate(current)
+
+    for _ in range(n_iter):
+        forward(start_node)
+    policy_count = torch.LongTensor([child.n_visit for child in start_node.children])
+    sampled_node = start_node.children[
+        torch.multinomial(torch.pow(policy_count, (1 / temperature)).float(), 1).item()
+    ]
+    # TODO Maybe just iterate through children once, more efficient?
+    # Currently doing it twice due to output difference, just do forloop instead?
+    return (
+        sum([child.action * child.n_visit for child in start_node.children]),
+        sampled_node,
+    )
+
+
+def self_play(
+    game: GameClassType, net: torch.nn.Module, n_mcts_iter: int = 1600,
+) -> List[tuple]:
+    pos = game.get_init_board()
+    temperature = 0.0001
+    data = []
+    is_terminal, res = game.is_terminal(pos)
```

Comment:
I think this has to be changed to reflect how Game works now

Reply:
Yup. I'll either just stash non-used code for now or updated it